### PR TITLE
feat: Quick order

### LINKED ISSUE #11817: [MASTER] Quick Order
# Quick Order

## What is Quick Order

A feature that allows customers to add multiple products to active their cart. https://help.sap.com/viewer/4c33bf189ab9409e84e589295c36d96e/2011/en-US/caf95981aa174660b3faf839a9dddbef.html?q=quick%20order

## Overview

**requirements**: https://cxwiki.sap.com/pages/viewpage.action?spaceKey=spar&title=B2B+Quick+Order+SoT+Nov2020
**mocks**: https://zpl.io/2ZZqWWl
**github labels**: `quick-order`. 

## EPIC branch

All tickets should be merged towards `epic/quick-order`. Final QA will happen in this branch before merging into the 3.x branch to make sure nothing is missed.

- Progress in the epic branch PR: #11872 

## IMPORTANT

1. Keep in mind accessibility
2. Do not forget about having a threat modeling session
3. Make sure the implemented CSS for the given component supports `Directionality`
4. Be careful of deprecations and adjust accordingly

More information can be found here https://sap.github.io/spartacus-docs/definition-of-done/

## Sample data

- [x] introduce Quick Order sample-data through impex (GH-11827)

## Implementation

1. creation of a minimal feature-lib
- [x] implement feature-lib for quick order (GH-11818)

2. create core logic
- [x] implement functionalities for **adding** the SKUs (GH-11821)
- [x] implement functionalities for **listing** the SKUs (GH-11822)
- [x] implement functionalities for **moving** the SKUs to active cart (GH-11823)
- [x] implement functionalities for cart form (GH-12697)

3. quick order page with the functionality
- [x] implement UI/UX for the form (GH-11825)
- [x] implement UI/UX for the list (GH-11826)

4. quick order cart form the functionality
- [x] implement UI/UX for the form (GH-12696)

5. quick order unit tests
- [x]  implementation unit tests for the form
- [x]  implementation unit tests for the list
- [x] implementation unit test for cart form

6. quick order e2es
- [x] implement quick order e2e (GH-11819)
- [x] implement accessibility e2e for quick order (GH-11820)

7. quick order doc
- [x] implement quick order docs (GH-11824)

## FUTURE / STRETCH
- [ ] implement suggestions based on SKU (GH-12886)
- [ ] implement undo of a deletion (GH-12883)
- [ ] implement search by name (GH-12884)
- [ ] implement import a CSV file that adds products to the Quick Order page (GH-12885)
- [ ] implements loading the quick order form with items from a saved cart (GH-12882)
- [ ] implement moving a product out of quick order to a saved cart (GH-12880)
- [ ] implement saving of the products to a saved cart instead of adding all to cart (GH-12879)

### PR DESCRIPTION
**Pull request by dydome**
closes #11817

Master ticket: #11817

### CODE REVIEW

File: projects/schematics/src/shared/constants.ts [hunk 1]
```diff
@@ -1075,6 +1075,8 @@ export const CLI_ASM_FEATURE = 'ASM';
 
 export const CLI_CART_SAVED_CART_FEATURE = 'Saved-Cart';
 
+export const CLI_CART_QUICK_ORDER_FEATURE = 'Quick-Order';
```
Comment (marlass) [inline projects/schematics/src/shared/constants.ts:1079]: The feature should also be added in schema.json in add-spartacus directory.

It is used to list all features to install when running `ng add @spartacus/schematics`

File: feature-libs/cart/schematics/add-cart/__snapshots__/index_spec.ts.snap [hunk 2]
```diff
@@ -1,5 +1,127 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`Spartacus Cart schematics: ng-add Quick Order feature eager loading should import appropriate modules 1`] = `""`;
+
+exports[`Spartacus Cart schematics: ng-add Quick Order feature general setup should add the feature using the lazy loading syntax 1`] = `""`;
```
Comment (marlass) [inline feature-libs/cart/schematics/add-cart/__snapshots__/index_spec.ts.snap]: These snapshots definitely should not be empty. Maybe you are snapshotting the wrong file?

File: feature-libs/cart/schematics/constants.ts [hunk 3]
```diff
@@ -13,3 +13,17 @@ export const SPARTACUS_SAVED_CART_ASSETS = `${SPARTACUS_SAVED_CART}/assets`;
 export const SAVED_CART_TRANSLATIONS = 'savedCartTranslations';
 export const SAVED_CART_TRANSLATION_CHUNKS_CONFIG =
   'savedCartTranslationChunksConfig';
+export const CLI_SAVED_CART_FEATURE = 'Saved Cart (b2b feature)';
```
Comment (marlass) [inline feature-libs/cart/schematics/constants.ts]: Where it's used?

File: feature-libs/cart/schematics/constants.ts [hunk 4]
```diff
@@ -13,3 +13,17 @@ export const SPARTACUS_SAVED_CART_ASSETS = `${SPARTACUS_SAVED_CART}/assets`;
...
+
+export const CART_QUICK_ORDER_MODULE_NAME = 'CartQuickOrder';
+export const CART_QUICK_ORDER_FEATURE_NAME_CONSTANT =
```
Comment (marlass) [inline feature-libs/cart/schematics/constants.ts:18]: Where it's used?

File: projects/storefrontstyles/scss/layout/page-templates/_cart-page.scss [hunk 5]
```diff
@@ -35,6 +35,10 @@
         max-width: 50%;
       }
 
+      cx-cart-quick-form {
```
Comment (marlass) [inline projects/storefrontstyles/scss/layout/page-templates/_cart-page.scss]: Why we need this file here instead of cart lib styles?

File: projects/core/src/model/cms.model.ts [hunk 6]
```diff
@@ -175,3 +175,7 @@ export interface CmsProductFacetNavigationComponent extends CmsComponent {
 export interface CmsAddToCartComponent extends CmsComponent {
   inventoryDisplay?: boolean;
 }
+
+export interface CmsQuickOrderComponent extends CmsComponent {
```
Comment (marlass) [inline projects/core/src/model/cms.model.ts]: Can we move this interface to the quick order lib code?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+import { Observable, Subscription } from 'rxjs';
+import { map, switchMap, tap } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-cart-quick-form',
+  templateUrl: './cart-quick-form.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CartQuickFormComponent implements OnInit, OnDestroy {
+  orderForm: FormGroup;
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Can we rename to `quickOrderForm`?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+  ngOnInit(): void {
+    this.cart$ = this.activeCartService.getActiveCartId().pipe(
+      tap((activeCardId: string) => (this.cartId = activeCardId)),
+      switchMap(() => this.activeCartService.getActive())
+    );
+
+    this.cartIsLoading$ = this.activeCartService
+      .isStable()
+      .pipe(map((loaded) => !loaded));
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Why not set the values directly when we declare properties?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    if (productCode && quantity) {
+      this.activeCartService.addEntry(productCode, quantity);
+    }
+  }
+
+  protected buildForm(): void {
+    this.orderForm = this.formBuilder.group({
+      productCode: ['', [Validators.required]],
+      quantity: [1, [Validators.required]],
+    });
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Can't we just assign the value when we declare property?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  protected watchAddEntrySuccessEvent(): void {
+    this.subscription.add(
+      this.eventService
+        .get(CartAddEntrySuccessEvent)
+        .subscribe((data: CartAddEntrySuccessEvent) => {
+          let key = 'quickOrderCartForm.stockLevelReached';
+          let productTranslation;
+          let messageType = GlobalMessageType.MSG_TYPE_WARNING;
+
+          if (data.quantityAdded && 0 < data.quantityAdded) {
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]:
```suggestion
          if (data.quantityAdded > 0) {
```

Can we simplify this (Am I missing something)?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+@Component({
+  selector: 'cx-cart-quick-form',
+  templateUrl: './cart-quick-form.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CartQuickFormComponent implements OnInit, OnDestroy {
+  orderForm: FormGroup;
+  cartIsLoading$: Observable<boolean>;
+  cart$: Observable<Cart>;
+  cartId: string;
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Is it used?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (productCode && quantity) {
+      this.activeCartService.addEntry(productCode, quantity);
+    }
+  }
+
+  protected buildForm(): void {
+    this.orderForm = this.formBuilder.group({
+      productCode: ['', [Validators.required]],
+      quantity: [1, [Validators.required]],
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Maybe use min value for initial value?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.module.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+      cmsComponents: {
+        CartQuickOrderFormComponent: {
+          component: CartQuickFormComponent,
+        },
+      },
+    }),
+  ],
+  declarations: [CartQuickFormComponent, CartQuickFormComponent],
+  exports: [CartQuickFormComponent],
+  entryComponents: [CartQuickFormComponent],
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.module.ts]: entryComponent are no longer needed
```suggestion
```

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.module.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  providers: [
+    provideDefaultConfig(<CmsConfig>{
+      cmsComponents: {
+        CartQuickOrderFormComponent: {
+          component: CartQuickFormComponent,
+        },
+      },
+    }),
+  ],
+  declarations: [CartQuickFormComponent, CartQuickFormComponent],
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.module.ts]: I guess we want to be sure it's declared ;)

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.module.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  exports: [CartQuickFormComponent],
+  entryComponents: [CartQuickFormComponent],
+})
+export class CartQuickFormModule {}
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.module.ts]: Can we reference the full feature name in the module name? Without order the quick reads weird

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+} from '@spartacus/core';
+import { Observable, Subscription } from 'rxjs';
+import { map, switchMap, tap } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-cart-quick-form',
+  templateUrl: './cart-quick-form.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CartQuickFormComponent implements OnInit, OnDestroy {
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Can we rename to `CartQuickOrderFormComponent`?

File: feature-libs/cart/quick-order/components/quick-order/form/quick-order-form.component.ts [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        (error: HttpErrorResponse) => {
+          this.globalMessageService.add(
+            error.error.errors[0].message,
+            GlobalMessageType.MSG_TYPE_ERROR
+          );
+        }
+      );
+  }
+
+  protected build() {
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/quick-order/form/quick-order-form.component.ts]: How about `buildForm`?

File: feature-libs/cart/quick-order/components/quick-order/quick-order.component.ts [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+  ) {}
+
+  ngOnInit(): void {
+    this.cartId$ = this.activeCartService.getActiveCartId();
+    this.entries$ = this.quickOrderService.getEntries();
+    this.quickOrderStatePersistenceService.initSync();
+    this.watchCartAddEntryEvents();
+  }
+
+  get errors(): any[] {
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/quick-order/quick-order.component.ts]: Avoid `any` and set the proper type

File: feature-libs/cart/quick-order/components/quick-order/quick-order.component.ts [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+
+    this.subscription.add(watchCartAddEntrySuccessEvent);
+    this.subscription.add(watchCartAddEntryFailEvent);
+  }
+
+  protected addError(
+    error: CartAddEntrySuccessEvent | CartAddEntryFailEvent
+  ): void {
+    this.cartErrors.push(error);
+  }
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/quick-order/quick-order.component.ts]: I feel like this whole logic should be encapsulated in other place (service).

We should start listening to events just before we start adding them to cart and we should stop listening once we complete the process. Listening for the whole lifetime of the application could cause some unexpected behaviour in this component.

File: feature-libs/cart/quick-order/components/quick-order/table/item/quick-order-item.component.ts [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+  removeEntry(): void {
+    this.quickOrderService.removeEntry(this.index);
+  }
+
+  protected watchProductAdd(): Subscription {
+    return this.quickOrderService.getProductAdded().subscribe(() => {
+      this.quantityControl = new FormControl(this.entry.quantity);
+      this.cd.detectChanges();
+    });
+  }
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/quick-order/table/item/quick-order-item.component.ts]: I feel like it's extra work (one entry changes, but it's run for every instance of this component). Changes in `@Input` are not enough to handle value of an item being changed?

File: feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        storageType: StorageSyncType.SESSION_STORAGE,
+        onRead: (state) => this.onRead(state),
+      })
+    );
+  }
+
+  /**
+   * Initializes the synchronization between state and browser storage.
+   */
+  getProductsList(): any {
```
Comment (marlass) [inline feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts]: Fix types in this service

File: projects/storefrontapp/src/app/spartacus/features/quick-order-feature.module.ts [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+} from '@spartacus/cart/quick-order/assets';
+import { QuickOrderRootModule } from '@spartacus/cart/quick-order/root';
+import { provideConfig } from '@spartacus/core';
+
+@NgModule({
+  imports: [QuickOrderRootModule],
+  providers: [
+    provideConfig({
+      featureModules: {
+        cartQuickOrder: {
```
Comment (marlass) [inline projects/storefrontapp/src/app/spartacus/features/quick-order-feature.module.ts]: use feature name variable you created for that

File: feature-libs/cart/quick-order/root/quick-order-root.module.ts [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+import {
+  provideDefaultConfig,
+  provideDefaultConfigFactory,
+  RoutingConfig,
+} from '@spartacus/core';
+
+export function defaultQuickOrderComponentsConfig() {
+  const config = {
+    featureModules: {
+      cartQuickOrder: {
```
Comment (marlass) [inline feature-libs/cart/quick-order/root/quick-order-root.module.ts]: use feature name you created for that

File: projects/core/src/occ/occ-models/occ-endpoints.model.ts [hunk 24]
```diff
@@ -658,5 +658,11 @@ export interface OccEndpoints {
    * @member {string}
    */
   saveCart?: string | OccEndpoint;
+  /**
+   * Endpoint for add quick order to active cart
+   *
+   * @member {string}
+   */
+  addToCart?: string | OccEndpoint;
```
Comment (marlass) [inline projects/core/src/occ/occ-models/occ-endpoints.model.ts]: Why it's added?

**Review by marlass — CHANGES_REQUESTED** [review]

File: projects/core/src/occ/occ-models/occ-endpoints.model.ts (hunk 24 above)
Comment (dydome) [reply projects/core/src/occ/occ-models/occ-endpoints.model.ts]: no needed, the old version required this part of code - removing

File: feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts (hunk 21 above)
Comment (dydome) [reply feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts]: unused method

File: projects/core/src/model/cms.model.ts (hunk 6 above)
Comment (dydome) [reply projects/core/src/model/cms.model.ts]: I was advised by Brian to put it here to keep everything in one place, should I move it or I can keep it here?

File: feature-libs/cart/schematics/constants.ts (hunk 4 above)
Comment (dydome) [reply feature-libs/cart/schematics/constants.ts:18]: in ...\feature-libs\cart\schematics\add-cart\index.ts - 104 line

File: feature-libs/cart/schematics/constants.ts (hunk 3 above)
Comment (dydome) [reply feature-libs/cart/schematics/constants.ts]: this is for saved cart but I removed it and also part for quick order

File: feature-libs/cart/quick-order/components/quick-order/table/item/quick-order-item.component.ts (hunk 20 above)
Comment (dydome) [reply feature-libs/cart/quick-order/components/quick-order/table/item/quick-order-item.component.ts]: yes, they were not enough, but I refactor it a little bit to verify product code

File: report.20210818.091334.24348.0.001.json [hunk 25]
```diff
@@ -0,0 +1,783 @@
+
+{
```
Comment (marlass) [inline report.20210818.091334.24348.0.001.json]: I don't think this file should've been commited

File: projects/storefrontstyles/scss/layout/page-templates/_cart-page.scss [hunk 26]
```diff
@@ -48,6 +48,10 @@
         max-width: none;
       }
 
+      cx-cart-quick-order-form {
```
Comment (marlass) [inline projects/storefrontstyles/scss/layout/page-templates/_cart-page.scss]: This one should also be moved to feature lib styles

File: projects/storefrontstyles/package.json [hunk 27]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@spartacus/styles",
-  "version": "4.0.0",
+  "version": "4.0.1",
```
Comment (marlass) [inline projects/storefrontstyles/package.json]: Revert this change

File: projects/storefrontapp/src/environments/environment.ts [hunk 28]
```diff
@@ -15,10 +15,12 @@ import { Environment } from './models/environment.model';
...
 export const environment: Environment = {
   production: false,
-  occBaseUrl: buildProcess.env.CX_BASE_URL,
+  occBaseUrl: 'https://spartacus-dev3.eastus.cloudapp.azure.com:9002/',
+  // occBaseUrl: buildProcess.env.CX_BASE_URL,
   occApiPrefix: '/occ/v2/',
   cds: buildProcess.env.CX_CDS ?? false,
-  b2b: buildProcess.env.CX_B2B ?? false,
+  // b2b: buildProcess.env.CX_B2B ?? false,
+  b2b: true,
```
Comment (marlass) [inline projects/storefrontapp/src/environments/environment.ts]: Quick order is b2b only feature?

File: projects/schematics/package.json [hunk 29]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@spartacus/schematics",
-  "version": "4.0.0",
+  "version": "4.0.1",
```
Comment (marlass) [inline projects/schematics/package.json]: This change should also be reverted

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+    protected eventService: EventService,
+    protected formBuilder: FormBuilder,
+    protected globalMessageService: GlobalMessageService
+  ) {}
+
+  ngOnInit(): void {
+    this.buildForm();
+    this.watchQuantityChange();
+    this.watchAddEntrySuccessEvent();
+    this.watchAddEntryFailEvent();
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Can we do the same thing we did with addToCart and entries that we only listen to those events when we submit form and we unsubscribe once it completes?

File: feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+} from '@spartacus/core';
+import { Observable, Subscription } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-cart-quick-order-form',
+  templateUrl: './cart-quick-form.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CartQuickOrderFormComponent implements OnInit, OnDestroy {
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/cart-quick-form/cart-quick-form.component.ts]: Filenames should also be updated to new name

File: feature-libs/cart/quick-order/components/quick-order/quick-order.component.ts [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+          (!!cartEvent.quantityAdded &&
+            cartEvent.quantityAdded < cartEvent.quantity)
+        ) {
+          this.addError(cartEvent);
+        }
+      });
+
+    this.cartEventsSubscription = new Subscription();
+    this.cartEventsSubscription.add(watchCartAddEntrySuccessEvent);
+  }
```
Comment (marlass) [inline feature-libs/cart/quick-order/components/quick-order/quick-order.component.ts]: I expected this to be part of addToCart method in the service

File: feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+        onRead: (state) => this.onRead(state),
+      })
+    );
+  }
+
+  /**
+   * Function called on each browser storage read.
+   * Used to update state from browser -> state.
+   */
+  protected onRead(state: any): void {
```
Comment (marlass) [inline feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts]: Change any to proper type

File: feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+    this.quickOrderService.loadEntries(state);
+  }
+
+  /**
+   * Reads synchronously state from storage and returns it.
+   */
+  protected readStateFromStorage() {
+    return this.statePersistenceService.readStateFromStorage<any>({
```
Comment (marlass) [inline feature-libs/cart/quick-order/core/services/quick-order-state-persistance.service.ts]: Use proper type here

File: projects/storefrontapp/src/environments/environment.ts (hunk 28 above)
Comment (dydome) [reply projects/storefrontapp/src/environments/environment.ts]: is b2b and b2c feature

**Review by marlass — APPROVED** [review]
